sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -68,7 +68,7 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] {
    * generating case class `equals` methods
    */
   def fastEquals(other: TreeNode[_]): Boolean = {
-    this.eq(other) || this == other
+    this.eq(other)// || this == other
```

Comment:
I thought that also before I made this change, but I don't think that the strong reason we should stop this change.

For most of cases in the code, we returns the same references by well-design the `rules` for a `TreeNode` object, and if the code still keep creating the identical objects(`.equals` returns `true`) in its rule for every iteration, even unnecessary, can this be considered as a bug of the user code? 

I think it will be the responsibility for user code to decide whether `TreeNode` object substitutions should be taken (via creating new instance), as user code always knows when a object substitution needed, right? That's also give more freedom for user code to define the `.equals()` in a semantic way for `TreeNode` object.